reps/2022-04-01-resource-control.md

```diff
@@ -0,0 +1,167 @@
+## Summary
+### General Motivation
+
+Currently, we don't control and limit the resource usage of worker processes, except running the worker in a container (see the `container` part in the [doc](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference)). In most of the scenarios, the container is unnecessary, but resource control is necessary for isolation.
+
+[Control groups](https://man7.org/linux/man-pages/man7/cgroups.7.html), usually referred to as cgroups, are a Linux kernel feature which allow processes to be organized into hierarchical groups whose usage of various types of resources can then be limited and monitored.
+
+So, the goal of this proposal is to achieve resource control for worker processes by cgroup in Linux.
+
+### Should this change be within `ray` or outside?
+
+These changes would be within Ray core.
+
+## Stewardship
+### Required Reviewers
+@ericl, @edoakes, @simon-mo, @chenk008 @raulchen 
+
+### Shepherd of the Proposal (should be a senior committer)
+@ericl
+
+## Design and Architecture
+
+### Cluster level API
+We should add some new system configs for the resource control.
+- worker_resource_control_method: Set to "cgroup" by default.
+- cgroup_manager: Set to "cgroupfs" by default. We should also support `systemd`.
+- cgroup_mount_path: Set to `/sys/fs/cgroup/` by default.
+- cgroup_unified_hierarchy: Whether to use cgroup v2 or not. Set to `False` by default.
+- cgroup_use_cpuset: Whether to use cpuset. Set to `False` by default.
+
+### User level API
+#### Simple API using a flag
+```python
+    runtime_env = {
+        "enable_resource_control": True,
```

Comment:
Is there any reason this cannot be implemented using runtime_env plugins? We should consider whether to extend the plugin mechanism to support this, if this is not possible, rather than special casing cgroups.

The goal of runtime env plugins is exactly to enable supporting these kind of complex use cases in a customizable way.

Reply:
In the current runtime env plugin mechanism design, a plugin must implement a whole runtime env type (e.g., pip, conda, etc).
Cgroup is different here. Because it is more like a config of runtime env, rather than a new runtime env type. Because it can be used together with other types. 
So in order to pluginize cgroup, we may need a new plugin mechanism that can extend the the current plugin mechanism. This seems weird. and I'm not sure if such a new plugin mechanism can be re-used for other features. 
So I think maybe we can consider pluginizing it later if we have a second similar requirement.

Reply:
How to achieve the logic of deleting cgroups in plugin?

Reply:
Sorry, let me explain what I said above.

with the current plugin mechanism, it's hard to implement cgroup deletion. As I said here https://github.com/ray-project/enhancements/pull/7#discussion_r847002806, the cgroups are one-off. It couldn't be shared and reused. So, the logic should be that: Create cgroup -> start worker -> worker dies -> delete cgroup. For example, if the cgroup dir is `/sys/fs/cgroup/{worker_id}`, we can't reuse the reference counting logic to delete this dir because we don't provide the `worker_id` when we decrease the reference by the PRC from raylet to agent.